### ISSUE TITLE
 Loss calculation bug in BartForCausalLM

### ISSUE THREAD

**Issue by willxxy** [huggingface/transformers#30105]
### System Info

transformers==4.39.1
python==3.8.18

### Who can help?

@ArthurZucker and @younesbelkada

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

The gist of the issue is found in this [issue](https://github.com/huggingface/transformers/issues/27517), where the primary problem is the way that the loss is being calculated in the BartForCausalLM module in modeling_bart.py file.

The current implementation of the loss is like so 

```
loss = None
if labels is not None:
     labels = labels.to(logits.device)
     loss_fct = CrossEntropyLoss()
     loss = loss_fct(logits.view(-1, self.config.vocab_size), labels.view(-1))
```

This I believe is the mlm loss calculation and not aligned with the causallm loss calculation.

### Expected behavior

I am not entirely sure but I wanted confirmation from the quick fix that was shown in [issue](https://github.com/huggingface/transformers/issues/27517).

It seems like in the forward function of BartForCausalLM in modeling_bart.py instead of calculating the loss like this 

```
loss = None
if labels is not None:
     labels = labels.to(logits.device)
     loss_fct = CrossEntropyLoss()
     loss = loss_fct(logits.view(-1, self.config.vocab_size), labels.view(-1))
```

they switched it out with this

```
loss = None
if labels is not None:
    labels = labels.to(logits.device)

    # Shift so that tokens < n predict n
    shift_logits = logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()

    loss_fct = CrossEntropyLoss()
    loss = loss_fct(shift_logits.view(-1, self.config.vocab_size), shift_labels.view(-1))

```

Additionally, the labels in the modified causal loss case would be similar to that of Llama where the labels would be the same as the inputs since we are shifting them in the loss calculation.
I wanted to confirm whether this is the correct way to do this. 

Thank you!

**Comment by ArthurZucker** [comment]
Llama is probably up to date, but changing this today would be breaking 🤗

**Comment by willxxy** [comment]
Ok got it. Thank you!
